#22430 implement TlsModule materialization phase

jrudolph opened the pull request:
Preview to be able to discuss bits of wip code.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/io/TlsModule.scala:
```diff
@@ -29,7 +30,7 @@ private[stream] final case class TlsModule(plainIn: Inlet[SslTlsOutbound], plain
   override def toString: String = f"TlsModule($closing) [${System.identityHashCode(this)}%08x]"
 
-  override private[stream] def traversalBuilder = TraversalBuilder.atomic(this)
+  override private[stream] def traversalBuilder = TraversalBuilder.atomic(this).makeIsland(TlsModuleIslandTag)
```
I would expect that this ensures that above we only ever get a `TlsModule` as the module in `materializeAtomic`.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala:
```diff
@@ -723,4 +730,73 @@ final class ProcessorModulePhase(materializer: PhasedFusingActorMaterializer) ex
+  def materializeAtomic(mod: AtomicModule[Shape, Any], attributes: Attributes): (MaterializedInfo, Any) = {
+    val es = materializer.effectiveSettings(attributes)
+    println(mod)
+    val tls = mod.asInstanceOf[TlsModule]
```
For the `TlsSpec` this fails for me with a `ClassCastException`. The module printed above is a

```
GraphStage(Map(<function1>)) [4313baad]
```

How can that be, @drewhk?

drewhk replied:
Try running a simple test with the `Debug` flag in the materializer on. I suspect that the layout is wrong and very likely the same bug I am fixing.

jrudolph replied:
Here's the layout:

```
--- Materializing layout:
 | push attr Attributes(List(Name(headSink), InputBuffer(1,1)))
 |  |  | materialize GraphStage(HeadOptionStage) [28ac7cef] []
 |  | transform mat
 | pop attr
 | push attr Attributes(List())
 |  | materialize GraphStage(DropWhile) [52dab3fe] [-1]
 |  | pop attr
 |  | pop mat
 |  | push attr Attributes(List())
 |  | materialize GraphStage(akka.stream.io.TlsSpec$Timeout@7c40e710) [65510559] [-1]
 |  | pop attr
 |  | pop mat
 |  | push attr Attributes(List())
 |  | materialize GraphStage(Scan) [3740b773] [-1]
 |  | pop attr
 |  | pop mat
 |  | push attr Attributes(List())
 |  | materialize GraphStage(Collect) [3002a726] [-1]
 |  | pop attr
 |  | pop mat
 |  | push attr Attributes(List())
 |  | materialize GraphStage(Map(class akka.stream.io.TlsSpec$$anonfun$1$$anonfun$7)) [5d1407b2] [-1]
 |  | pop mat
 |  |  | push NotUsed
 |  | pop attr
 |  | pop mat
 |  | push attr Attributes(List())
 |  | materialize GraphStage(akka.stream.io.TlsSpec$$anonfun$1$$anonfun$apply$mcV$sp$6$$anonfun$apply$6$$anonfun$apply$7$$anonfun$apply$8$$anon$1@14ded3e9) [26970367] [-1]
 |  | pop attr
 |  | pop mat
 |  | push attr Attributes(List())
 |  | enter island akka.stream.impl.TlsModuleIslandTag$@6c3b70ec
 |  |  |  | materialize TlsModule(IgnoreComplete) [31406dc3] [3, -1]
 |  |  |  | enter island akka.stream.impl.TlsModuleIslandTag$@6c3b70ec
 |  |  |  |  | materialize TlsModule(IgnoreComplete) [4c14d5ce] [-1, 3]
 |  |  |  | pop mat
 |  |  |  | push attr Attributes(List())
 |  |  |  |  | materialize GraphStage(Map(class akka.stream.io.TlsSpec$$anonfun$1$PayloadScenario$1$$anonfun$flow$1)) [52066b10] [-2]
 |  |  |  |  | pop mat
 |  |  |  |  |  | push NotUsed
 |  |  |  |  | pop attr
 |  |  |  |  | pop mat
 |  |  |  |  | push attr Attributes(List())
 |  |  |  |  | materialize GraphStage(Map(class akka.stream.io.TlsSpec$$anonfun$1$$anonfun$7)) [5d1407b2] [-1]
 |  |  |  |  | pop mat
 |  |  |  | push NotUsed
 |  |  | pop mat
 |  | pop attr
 |  | pop mat
 |  | push attr Attributes(List(Name(iterableSource)))
 |  |  | push attr Attributes(List())
 |  |  | materialize GraphStage(StatefulMapConcat) [3025ab22] [-6]
 |  |  | pop attr
 |  |  | pop mat
 |  | materialize GraphStage(SingleSource(List(SendBytes(ByteString(104, 101, 108, 108, 111)), SendBytes(ByteString()), SendBytes(ByteString(32, 119, 111, 114, 108, 100))))) [0239e690] [-1]
 | pop attr
pop mat
```

I'll try to simplify the flow tomorrow to make it simpler to debug.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala:
```diff
@@ -232,7 +237,7 @@ class IslandTracking(
         if (Debug) println(s"    cross island forward wiring from port ${forwardWire.from} wired to local slot = $localInSlot")
         val publisher = forwardWire.phase.createPublisher(forwardWire.from, forwardWire.outStage)
-        currentPhase.takePublisher(localInSlot, publisher)
+        currentPhase.takePublisher(localInSlot, publisher, logic)
```
that is the wrong logic there. There is a reason why the logic cannot be passed here.

drewhk replied:
Ah, no you are right it can be passed here. It is createPublisher that cannot take a logic, sorry.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala:
```diff
@@ -723,4 +730,73 @@ final class ProcessorModulePhase(materializer: PhasedFusingActorMaterializer) ex
+    val tls = mod.asInstanceOf[TlsModule]
+
+    val props = TLSActor.props(materializer.settings, tls.createSSLEngine, tls.verifySession, tls.closing)
+    val impl = materializer.actorOf(props, s"tls-${nextId()}", es.dispatcher)
```
I think the ID generation should be moved into the `actorOf` method instead. This is related to what @patriknw is working on, so maybe coordinate on this?

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala:
```diff
@@ -508,7 +513,9 @@ trait PhaseIsland[M] {
-  def takePublisher(slot: Int, publisher: Publisher[Any]): Unit
+  def takePublisher(slot: Int, publisher: Publisher[Any], logic: M): Unit = takePublisher(slot, publisher)
+
+  protected def takePublisher(slot: Int, publisher: Publisher[Any]): Unit
```
put some docs on those why we have two?

drewhk replied:
We should not have two in the first place.

jrudolph replied:
Yep, that's only so that I don't have to do wide-reaching changes just for trying things out.

ktoso replied:
ok :)

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala:
```diff
@@ -276,7 +281,7 @@ class IslandTracking(
           if (Debug) println(s"    cross-island wiring to local slot $localInSlot in target island")
           val publisher = currentPhase.createPublisher(out, logic)
-          targetSegment.phase.takePublisher(localInSlot, publisher)
+          targetSegment.phase.takePublisher(localInSlot, publisher, logic)
```
This is wrong. We are wiring back to a module we already visited. This logic you pass here is not that logic. And now I remember why it does not take a `logic` as parameter and why it is not passed in L240. If the island needs the logic it has to keep track of it itself, but it should not impose the tracking cost to islands that don't need it.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala:
```diff
@@ -723,4 +730,73 @@ final class ProcessorModulePhase(materializer: PhasedFusingActorMaterializer) ex
+  def name: String = "TlsModulePhase"
+
+  def materializeAtomic(mod: AtomicModule[Shape, Any], attributes: Attributes): (MaterializedInfo, Any) = {
+    val es = materializer.effectiveSettings(attributes)
```
the settings you get from the Phase factory callback already contains the effective settings, no need to calculate it again.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala:
```diff
@@ -723,4 +730,73 @@ final class ProcessorModulePhase(materializer: PhasedFusingActorMaterializer) ex
+    assignPort(tls.cipherIn, FanIn.SubInput[Any](impl, TLSActor.TransportIn))*/
+
+    //matVal.put(atomic, NotUsed)
+    (MaterializedInfo(
```
no need for this. Just store all this information in `var` fields

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala:
```diff
@@ -723,4 +730,73 @@ final class ProcessorModulePhase(materializer: PhasedFusingActorMaterializer) ex
+  protected def takePublisher(slot: Int, publisher: Publisher[Any]): Unit = ???
+
+  override def takePublisher( /*in: InPort, */ slot: Int, publisher: Publisher[Any], logic: MaterializedInfo): Unit = {
+    val in = logic.slots.find(_._1 == slot).get._2
```
no need, there are only two slots here, the two inputs. A simple `if` statement is enough as slot can be either 0 or 1 and it is deterministic which input port it means (the first or second).

jrudolph committed:
=str #22430 add TlsModulePhase

jrudolph commented:
Here's a little test I tried to see where wiring is broken:

```scala
object SimpleTlsTest extends App {
  implicit val system = ActorSystem()
  implicit val mat = ActorMaterializer()

  try {
    val tls = TLS(() ⇒ null: SSLEngine, null: TLSClosing)

    def full1 =
      tls join Flow[ByteString] join Flow[SslTlsInbound].map(_ ⇒ null: SslTlsOutbound)

    def full2 =
      Flow[SslTlsInbound].map(_ ⇒ null: SslTlsOutbound) join tls join Flow[ByteString]

    full2.run()

  } finally system.terminate()
}
```

`full1` fails because the `Identity` flow is inside the tls island. `full2` works.

`full1` layout:

```
--- Materializing layout:
 | push attr Attributes(List())
 | materialize GraphStage(Map(<function1>)) [61539c43] [1]
 | pop mat
 |  | push NotUsed
 | pop attr
 | pop mat
 | push attr Attributes(List())
enter island akka.stream.impl.TlsModuleIslandTag$@176f6ec1
 |  | materialize TlsModule(null) [256708aa] [2, -1]
 |  | push attr Attributes(List())
 |  | materialize GraphStage(Identity) [52fc6aa9] [-1]
 | pop mat
```

`full2` layout:

```
--- Materializing layout:
 | push attr Attributes(List())
 | materialize GraphStage(Identity) [518f91d3] [3]
 | pop attr
 | pop mat
 | push attr Attributes(List())
 | push attr Attributes(List())
 |  | materialize GraphStage(Map(<function1>)) [6a7abd08] [1]
 |  | pop mat
 | push NotUsed
 | enter island akka.stream.impl.TlsModuleIslandTag$@1c2f96ae
 |  | materialize TlsModule(null) [372bc57e] [-2, -1]
pop mat
```

drewhk commented:
Ok, I think it is good enough for now, make this mergeable and let's merge it. I know one of the issues at play here, fixing it soon.

jrudolph commented:
Rebased on top of current state (Patrik's actor name changes) and simplified. We can either merge now or when the traversal has been checked so we can see if the tests pass.

drewhk commented:
Well, it will not make more tests fail, so I see no reason for not merging
